extensions/pkg/controller/healthcheck/worker/nodes.go

```diff
@@ -0,0 +1,194 @@
+func (h *DefaultHealthChecker) SetLoggerSuffix(provider, extension string) {
+	h.logger = log.Log.WithName(fmt.Sprintf("%s-%s-healthcheck-nodes", provider, extension))
+}
+
+// DeepCopy clones the healthCheck struct by making a copy and returning the pointer to that new copy.
+func (h *DefaultHealthChecker) DeepCopy() healthcheck.HealthCheck {
+	copy := *h
+	return &copy
+}
+
+// Check executes the health check.
+func (h *DefaultHealthChecker) Check(ctx context.Context, request types.NamespacedName) (*healthcheck.SingleCheckResult, error) {
+	machineDeploymentList := &machinev1alpha1.MachineDeploymentList{}
+	if err := h.seedClient.List(ctx, machineDeploymentList, client.InNamespace(request.Namespace)); err != nil {
+		err := fmt.Errorf("nodes check failed. Failed to list machine deployments in namespace %s: %v", request.Namespace, err)
+		h.logger.Error(err, "Health check failed")
+		return nil, err
+	}
+
+	nodeList := &corev1.NodeList{}
+	if err := h.shootClient.List(ctx, nodeList); err != nil {
+		err := fmt.Errorf("nodes check failed. Failed to list shoot nodes: %v", err)
+		h.logger.Error(err, "Health check failed")
+		return nil, err
+	}
+
+	var (
+		readyNodes      int
+		registeredNodes = len(nodeList.Items)
+		desiredMachines = getDesiredMachineCount(machineDeploymentList.Items)
+	)
+
+	for _, node := range nodeList.Items {
+		if node.Spec.Unschedulable {
+			continue
+		}
+		for _, condition := range node.Status.Conditions {
+			if condition.Type == corev1.NodeReady && condition.Status == corev1.ConditionTrue {
+				readyNodes++
+			}
+		}
+	}
+
+	machineList := &machinev1alpha1.MachineList{}
+	if registeredNodes != desiredMachines || readyNodes != desiredMachines {
+		if err := h.seedClient.List(ctx, machineList, client.InNamespace(request.Namespace)); err != nil {
+			err := fmt.Errorf("nodes check failed. Failed to list machines in namespace %s: %v", request.Namespace, err)
+			h.logger.Error(err, "Health check failed")
+			return nil, err
+		}
+	}
+
+	// First check if the MachineDeployments report failed machines. If false then check if the MachineDeployments are
+	// "available". If false then check if there is a regular scale-up happening or if there are machines with an errorneuos
+	// phase. Only then check the other MachineDeployment conditions. As last check, check if there is a scale-down happening
+	// (e.g., in case of an rolling-update).
+
+	for _, deployment := range machineDeploymentList.Items {
+		for _, failedMachine := range deployment.Status.FailedMachines {
+			err := fmt.Errorf("machine %s failed: %s", failedMachine.Name, failedMachine.LastOperation.Description)
+			h.logger.Error(err, "Health check failed")
+			return &healthcheck.SingleCheckResult{
+				Status: gardencorev1beta1.ConditionFalse,
+				Detail: err.Error(),
+				Reason: "MachineDeploymentHasFailedMachines",
+				Codes:  gardencorev1beta1helper.DetermineErrorCodes(err),
+			}, nil
+		}
+
+		for _, condition := range deployment.Status.Conditions {
```

Comment:
Hmmm, you are executing the whole code below for every MachineDeployment, although it is not dependent on them.

Reply:
If I see it correctly, you are still making a call to  `checkMachineDeploymentsHealthy` and `checkNodesScalingDown` for every MachineDeployment in `machineDeploymentList` but those functions loop over all deployments/machines again for every call (so n^2 effort) and are independent of the current `deployment` here in this loop. Can we move these calls out of this loop to not make unnecessary duplicated checks with quadratic effort, or am I missing something?

Also, `checkNodesScalingUp` is now only called if the current `deployment` is not available, but still the func loops over all machines again and is not bound to the current `deployment`.
Can we enhance this check to bind it to a specific deployment or at least run it only once?

Reply:
Yes, indeed, nice catch, thanks again for the thorough review. Will adapt.